### ISSUE TITLE
In docstrings, the parameter type "container" has ambiguous meaning

### ISSUE THREAD

**Issue by Matthew-Signorotti** [networkx/networkx#5377]
The term "container" often refers to an object with a `__contains__` method, as in the `collections.abc.Container` abstract base class, or `typing.Container` in the case of type hints. However, multiple function docstrings state that some parameter is to accept containers of some type, when functionality besides a `__contains__` method is expected.

For example, in NetworkX 2.7.1, argument `ebunch_to_add` to `Graph.add_edges_from` is described as a "container of edges," but in the method body, it is directly accessed only on the line
```
        for e in ebunch_to_add:
```
`ebunch_to_add` is thus more precisely described as an "iterable of edges." If `ebunch_to_add` has a `__contains__` method but is not iterable, an exception is raised on this line.

Does "container" have an advantage over more precise, less ambiguous wording in describing the type of such parameters?

**Comment by rossbar** [comment]
> Does "container" have an advantage over more precise, less ambiguous wording in describing the type of such parameters?

No, I think switching to more precise language (i.e. iterable over container) would be an improvement.

**Comment by rossbar** [comment]
I'm sure there are many instances of this besides `add_edges_from` - it'd be good to try to fix them all at once. As a starting point, here is the result of `grep -rI "container of"`:

```
./networkx/classes/multigraph.py:        ebunch_to_add : container of edges
./networkx/classes/multigraph.py:        ebunch: list or container of edge tuples
./networkx/classes/function.py:        A container of nodes.
./networkx/classes/function.py:        A container of nodes.  A path will be constructed from
./networkx/classes/function.py:        A container of nodes.  A cycle will be constructed from
./networkx/classes/function.py:       A container of nodes that will be iterated through once (thus
./networkx/classes/function.py:    nbunch : node, container of nodes or None (for all nodes)
./networkx/classes/graph.py:            A container of nodes (list, dict, set, etc.).
./networkx/classes/graph.py:            A container of (node, attribute dict) tuples.
./networkx/classes/graph.py:            A container of nodes (list, dict, set, etc.).  If a node
./networkx/classes/graph.py:        ebunch_to_add : container of edges
./networkx/classes/graph.py:        ebunch_to_add : container of edges
./networkx/classes/graph.py:        ebunch: list or container of edge tuples
./networkx/classes/graph.py:            A container of nodes which will be iterated through once.
./networkx/classes/reportviews.py:    nbunch : node, container of nodes, or None meaning all nodes (default=None)
./networkx/classes/reportviews.py:    nbunch : node, container of nodes, or None meaning all nodes (default=None)
./networkx/classes/reportviews.py:    nbunch : container of nodes, node or None (default None)
./networkx/classes/reportviews.py:        nbunch : container of nodes, optional (default=None)
./networkx/classes/digraph.py:            A container of nodes (list, dict, set, etc.).
./networkx/classes/digraph.py:            A container of (node, attribute dict) tuples.
./networkx/classes/digraph.py:            A container of nodes (list, dict, set, etc.).  If a node
./networkx/classes/digraph.py:        ebunch_to_add : container of edges
./networkx/classes/digraph.py:        ebunch: list or container of edge tuples
./networkx/generators/classic.py:    n : int or iterable container of nodes
./networkx/generators/classic.py:        If n is a container of nodes, those nodes appear in the graph.
./networkx/generators/classic.py:    n : int or iterable container of nodes
./networkx/generators/classic.py:        If n is a container of nodes, those nodes appear in the graph.
./networkx/generators/classic.py:    n : int or iterable container of nodes (default = 0)
./networkx/generators/classic.py:        If n is a container of nodes, those nodes appear in the graph.
./networkx/generators/classic.py:    m, n : int or iterable container of nodes (default = 0)
./networkx/generators/lattice.py:    m, n : int or iterable container of nodes
./networkx/algorithms/distance_measures.py:    #    else:                      # assume v is a container of nodes
./networkx/algorithms/assortativity/mixing.py:    xy : list or container of two-tuples
./networkx/algorithms/bipartite/matching.py:    top_nodes : container of nodes
./networkx/algorithms/bipartite/basic.py:    The container of nodes passed as argument must contain all nodes
./networkx/algorithms/bipartite/basic.py:    The container of nodes passed as argument must contain all nodes
./networkx/algorithms/bipartite/cluster.py:        A container of nodes to use in computing the average.
./networkx/algorithms/bipartite/cluster.py:    The container of nodes passed to this function must contain all of the nodes
./networkx/algorithms/approximation/kcomponents.py:            A container of nodes.  The container will be iterated
./networkx/algorithms/simple_paths.py:    ignore_nodes : container of nodes
./networkx/algorithms/simple_paths.py:    ignore_edges : container of edges
./networkx/algorithms/simple_paths.py:    ignore_nodes : container of nodes
./networkx/algorithms/simple_paths.py:    ignore_edges : container of edges
./networkx/algorithms/reciprocity.py:    nodes : container of nodes, optional (default=whole graph)
./networkx/algorithms/cluster.py:    nodes : container of nodes, optional (default= all nodes in G)
./networkx/algorithms/cluster.py:    nodes : container of nodes, optional (default=all nodes in G)
./networkx/algorithms/cluster.py:    nodes : container of nodes, optional (default=all nodes in G)
./networkx/algorithms/cluster.py:    nodes : container of nodes, optional (default=all nodes in G)
./networkx/algorithms/cluster.py:    nodes : container of nodes, optional (default=all nodes in G)
./networkx/utils/decorators.py:    graph_types : container of strings
./networkx/utils/decorators.py:    """Decorator to allow number of nodes or container of nodes.
./doc/reference/glossary.rst:      An iteratable container of edge tuples like a list, iterator,
./doc/reference/glossary.rst:      An nbunch is a single node, container of nodes or `None` (representing
./doc/tutorial.rst:container of edge-tuples.  An edge-tuple can be a 2-tuple of nodes or a 3-tuple
./doc/tutorial.rst:a node, or an iterable container of nodes that is not itself a node in the
./examples/subclass/plot_antigraph.py:            A container of nodes.  The container will be iterated
```

I think a reasonable approach to fixing this issue is to go down the list and audit each instance individually. Some already say things like "an iterable container of nodes", which is closer to correct but still probably not *exactly* correct because I bet some of them accept generators (which are iterables but *not* containers).

**Comment by dschult** [comment]
I am sure that many /most of these are not quite wored correctly.
Is there a glossary of other source of correct wording for the different types of ...  well...  of ...  collections?  Anyway, objects that refer to other objects and might iterate, check membership, generate, index, slice, set operations, etc.

It's often hard to know exactly what term to put in there, so it'd be nice to have a source where we can look it up. :}
[and we could refer to that in our contributor guide]

**Comment by rossbar** [comment]
> Is there a glossary of other source of correct wording for the different types of ... well... of ... collections? Anyway, objects that refer to other objects and might iterate, check membership, generate, index, slice, set operations, etc.

The definitive resource is the `abc.collections` docs, particularly this table: https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes. I suspect the case highlighted by @Matthew-Signorotti in the OP is by far the most common case (the docs say "container" when they mean "iterable"), but I suspect there are also nuanced cases. For example, one might be tempted to replace "container" and "iterable" with "collection", which inherits from both; but many of the listed functions/methods accept generators, which are definitely *not* `Collection`s.

Long story short, this should be a relatively straightforward project, but it will require looking at each instance case-by-case (and a fairly good understanding of the Python collection abc's)!

**Comment by akshitasure12** [comment]
Have these changes been implemented in the respective files? If not, I'd be happy to take a look and work on it!

**Comment by VanitasCodes** [comment]
Hi @rossbar @dschult! I see PR #7991 was closed. I'd like to help with this issue.

Based on the discussion, I understand the approach should be:
1. Replace "container of" with more precise terminology (usually "iterable")
2. Consider what we want users to pass in, not just what technically works
3. Start with a small subset of files

I'll begin with `networkx/classes/graph.py` and submit a focused PR. If that approach looks good, I can continue with other files.

Does this sound reasonable?
